File: test/test.ReverseWebSocketUpstreamTransport.ts
import assert from "node:assert/strict";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "vitest";

import { LocalBrowserLauncher } from "../bridge/LocalBrowserLauncher.js";
import { ReverseWebSocketUpstreamTransport } from "../bridge/ReverseWebSocketUpstreamTransport.js";
import { ModCDPClient } from "../client/js/ModCDPClient.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const EXTENSION_PATH = path.resolve(HERE, "..", "dist", "extension");

test("reversews upstream config owns bind updates, wait timeout, and injector config", async () => {
  const transport = new ReverseWebSocketUpstreamTransport("127.0.0.1:29292", 10);
  assert.equal(transport.url, "ws://127.0.0.1:29292");
  assert.deepEqual(transport.getInjectorConfig(), { reverse_proxy_url: "ws://127.0.0.1:29292" });
  assert.equal(transport.update({ reversews_bind: "127.0.0.1:29293", reversews_wait_timeout_ms: 5 }), transport);
  assert.equal(transport.url, "ws://127.0.0.1:29293");
  assert.deepEqual(transport.getInjectorConfig(), { reverse_proxy_url: "ws://127.0.0.1:29293" });
  await assert.rejects(() => transport.waitForPeer(), /Timed out waiting 5ms/);
});

test("reversews upstream close rejects pending peer waits", async () => {
  const reverse_port = await LocalBrowserLauncher.freePort();
  const transport = new ReverseWebSocketUpstreamTransport(`127.0.0.1:${reverse_port}`, 5_000);
  const pending = transport.waitForPeer();

  await transport.close();

  await assert.rejects(
    () => pending,
    new RegExp(`Reverse websocket transport at ws://127\\.0\\.0\\.1:${reverse_port} closed before a peer connected`),
  );
});

test("reversews upstream accepts a real extension reverse connection and routes CDP through loopback", async () => {
  const reverse_port = await LocalBrowserLauncher.freePort();
  const reverse_bind = `127.0.0.1:${reverse_port}`;
  const reverse_url = `ws://${reverse_bind}`;
  const reverse = new ModCDPClient({
    launch: {
      mode: "local",
      options: { headless: true, sandbox: process.platform !== "linux" },
    },
    upstream: { mode: "reversews", reversews_bind: reverse_bind },
    extension: {
      mode: "auto",
      path: EXTENSION_PATH,
      service_worker_url_suffixes: ["/modcdp/service_worker.js"],
      trust_service_worker_target: true,
    },
    server: {
      routes: { "*.*": "loopback_cdp" },
    },
  });

  try {
    await reverse.connect();
    assert.equal(reverse.transport?.mode, "reversews");
    assert.equal(reverse.upstream_endpoint_kind, "modcdp_server");
    assert.equal(reverse.transport?.url, reverse_url);
    const version = (await reverse.send("Browser.getVersion")) as Record<string, unknown>;
    assert.equal(typeof version.product, "string");
  } finally {
    await reverse.close();
  }
}, 60_000);

File: client/python/tests/test_ReverseWebSocketUpstreamTransport.py
from __future__ import annotations

import socket
import unittest

from modcdp import ModCDPClient


class ReverseWebSocketUpstreamTransportTests(unittest.TestCase):
    def test_accepts_real_extension_reverse_connection_and_routes_cdp_through_loopback(self) -> None:
        reverse_port = _free_port()
        reverse_bind = f"127.0.0.1:{reverse_port}"
        cdp = ModCDPClient(
            launch={"mode": "local", "options": {"headless": True, "sandbox": False}},
            upstream={"mode": "reversews", "reversews_bind": reverse_bind},
            extension={
                "mode": "auto",
                "service_worker_url_suffixes": ["/modcdp/service_worker.js"],
                "trust_service_worker_target": True,
            },
            server={"routes": {"*.*": "loopback_cdp"}},
        )

        try:
            cdp.connect()
            self.assertEqual(cdp.transport.mode if cdp.transport else None, "reversews")
            self.assertEqual(cdp.upstream_endpoint_kind, "modcdp_server")
            version = cdp.send("Browser.getVersion")
            self.assertIsInstance(version["product"], str)
        finally:
            cdp.close()


def _free_port() -> int:
    sock = socket.socket()
    sock.bind(("127.0.0.1", 0))
    try:
        return int(sock.getsockname()[1])
    finally:
        sock.close()


if __name__ == "__main__":
    unittest.main()
